File: src/lib/api/preset.ts
// 预设组 v2 API 封装：与 src-tauri/src/commands/preset.rs 一一对应（参数 camelCase 由 Tauri 自动映射 snake_case）
import { invoke } from "@tauri-apps/api/core";
import type { ActivePreset, ResourceBinding } from "@/types/extension";
import type { ApplyPreview, PresetApplyResult, PresetRecord, RestoreResult } from "@/types/preset";

export async function listPresets(): Promise<PresetRecord[]> {
  return await invoke<PresetRecord[]>("list_presets");
}
export async function getPreset(presetId: string): Promise<PresetRecord | null> {
  return await invoke<PresetRecord | null>("get_preset", { presetId });
}
export async function createPreset(
  name: string,
  items: [string, string][],
  description?: string,
  scope?: string,
  boundTool?: string
): Promise<string> {
  return await invoke<string>("create_preset", { name, items, description, scope, boundTool });
}
export async function updatePreset(
  id: string,
  name: string,
  description: string,
  scope: string,
  boundTool: string | null,
  items: [string, string][]
): Promise<void> {
  return await invoke("update_preset", { id, name, description, scope, boundTool, items });
}
export async function deletePreset(presetId: string): Promise<void> {
  return await invoke("delete_preset", { presetId });
}
export async function applyPreset(presetId: string, toolId: string): Promise<PresetApplyResult> {
  return await invoke<PresetApplyResult>("apply_preset", { presetId, toolId });
}
// 工具级 deactivate 封装已退役（T8）：v2 预设×工具开关的「关」走 restore_preset
// 恢复默认，不再直呼 deactivate_preset（Rust 命令保留，子 Agent 级 deactivatePresetFromSubagent 仍用旧命令）
export async function restorePreset(toolId: string): Promise<RestoreResult> {
  return await invoke<RestoreResult>("restore_preset", { toolId });
}
export async function getActivePreset(toolId: string): Promise<string | null> {
  return await invoke<string | null>("get_active_preset", { toolId });
}
export async function listActivePresets(): Promise<ActivePreset[]> {
  return await invoke<ActivePreset[]>("list_active_presets");
}
export async function getToolActiveResources(toolId: string): Promise<[string, string, string][]> {
  return await invoke<[string, string, string][]>("get_tool_active_resources", { toolId });
}
export async function previewApplyPreset(presetId: string, toolId: string): Promise<ApplyPreview> {
  return await invoke<ApplyPreview>("preview_apply_preset", { presetId, toolId });
}
export async function applyPresetToSubagent(
  presetId: string,
  toolId: string,
  subAgentId: string
): Promise<PresetApplyResult> {
  return await invoke<PresetApplyResult>("apply_preset_to_subagent", {
    presetId,
    toolId,
    subAgentId,
  });
}
export async function deactivatePresetFromSubagent(
  presetId: string,
  toolId: string,
  subAgentId: string
): Promise<void> {
  return await invoke("deactivate_preset_from_subagent", { presetId, toolId, subAgentId });
}

// —— 资源独占绑定 / 常驻豁免 ——

export async function setResourceBinding(
  extensionId: string,
  exclusiveTools: string[],
  reason?: string
): Promise<void> {
  return await invoke("set_resource_binding", { extensionId, exclusiveTools, reason });
}
export async function listResourceBindings(): Promise<ResourceBinding[]> {
  return await invoke<ResourceBinding[]>("list_resource_bindings");
}
export async function deleteResourceBinding(extensionId: string): Promise<void> {
  return await invoke("delete_resource_binding", { extensionId });
}
export async function setToolResident(
  toolId: string,
  extensionId: string,
  resident: boolean
): Promise<void> {
  return await invoke("set_tool_resident", { toolId, extensionId, resident });
}
export async function listToolResidents(toolId: string): Promise<string[]> {
  return await invoke<string[]>("list_tool_residents", { toolId });
}

File: src-tauri/src/commands/preset.rs
// 预设组命令

use crate::database::{PresetRecord, ResourceBindingRecord};
use crate::services::preset::{ApplyPreview, PresetHealth, RestoreResult};

#[derive(serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PresetApplyResult {
    pub success_count: usize,
    pub failures: Vec<String>,
    pub conflicts: Vec<String>,
    pub stashed: Vec<String>,
    pub disabled: Vec<String>,
    pub restored_native: Vec<String>,
}

#[tauri::command]
pub fn create_preset(
    name: String,
    items: Vec<(String, String)>,
    description: Option<String>,
    scope: Option<String>,
    bound_tool: Option<String>,
) -> Result<String, String> {
    let scope = scope.unwrap_or_else(|| "universal".to_string());
    if scope == "tool" && bound_tool.is_none() {
        return Err("工具私有预设必须指定绑定工具".into());
    }
    crate::database::create_preset_with_meta(
        &name,
        &description.unwrap_or_default(),
        &scope,
        bound_tool.as_deref(),
        &items,
    )
}

#[tauri::command]
pub fn get_preset(preset_id: String) -> Option<PresetRecord> {
    crate::database::get_preset(&preset_id)
}

#[tauri::command]
pub fn update_preset(
    id: String,
    name: String,
    description: String,
    scope: String,
    bound_tool: Option<String>,
    items: Vec<(String, String)>,
) -> Result<(), String> {
    // Interfaces 契约：激活中的预设不可改（任一快照 active == id）——先恢复默认
    for tool in crate::adapter::TOOL_IDS {
        if let Some((Some(active), _)) = crate::database::get_base_snapshot(tool) {
            if active == id {
                return Err(format!(
                    "预设正在 {} 上激活，请先恢复默认再修改（PRESET_ACTIVE:{}）",
                    tool, id
                ));
            }
        }
    }
    if scope == "tool" && bound_tool.is_none() {
        return Err("工具私有预设必须指定绑定工具".into());
    }
    crate::database::update_preset(
        &id,
        &name,
        &description,
        &scope,
        bound_tool.as_deref(),
        &items,
    )
}

#[tauri::command]
pub fn delete_preset(preset_id: String) -> Result<(), String> {
    // spec §5.4：激活中的预设不可删——先恢复默认
    for tool in crate::adapter::TOOL_IDS {
        if let Some((Some(active), _)) = crate::database::get_base_snapshot(tool) {
            if active == preset_id {
                return Err(format!(
                    "预设正在 {} 上激活，请先恢复默认再删除（PRESET_ACTIVE:{}）",
                    tool, preset_id
                ));
            }
        }
    }
    crate::database::delete_preset(&preset_id)
}

#[tauri::command]
pub fn restore_preset(tool_id: String) -> Result<RestoreResult, String> {
    crate::services::tool_settings::ensure_tool_enabled(&tool_id)?;
    crate::services::preset::restore_tool(&tool_id)
}

/// 工具当前激活的预设（开关状态数据源；无激活 = None）
#[tauri::command]
pub fn get_active_preset(tool_id: String) -> Option<String> {
    crate::database::get_base_snapshot(&tool_id).and_then(|(active, _)| active)
}

/// 全工具激活预设（开关状态批量数据源，避免前端 N 次 invoke）
#[tauri::command]
pub fn list_active_presets() -> Vec<serde_json::Value> {
    crate::adapter::TOOL_IDS
        .iter()
        .filter_map(|tool| {
            crate::database::get_base_snapshot(tool)
                .and_then(|(active, _)| active)
                .map(|p| serde_json::json!({ "toolId": tool, "presetId": p }))
        })
        .collect()
}

/// 工具当前生效资源集合（FR-24「存为预设」预填数据源）
#[tauri::command]
pub fn get_tool_active_resources(tool_id: String) -> Vec<(String, String, String)> {
    crate::services::preset::snapshot::scan_tool_state(&tool_id)
        .into_iter()
        .map(|i| (i.extension_id, i.kind, i.origin))
        .collect()
}

/// 应用预览（确认弹窗数据源，dry-run）
#[tauri::command]
pub fn preview_apply_preset(preset_id: String, tool_id: String) -> Result<ApplyPreview, String> {
    // 与 apply_preset 对齐（评审裁决 4）：停用工具不该拿到执行不了的预览
    crate::services::tool_settings::ensure_tool_enabled(&tool_id)?;
    crate::services::preset::preview_apply(&preset_id, &tool_id)
}

#[tauri::command]
pub fn set_resource_binding(
    extension_id: String,
    exclusive_tools: Vec<String>,
    reason: Option<String>,
) -> Result<(), String> {
    crate::database::upsert_resource_binding(
        &extension_id,
        &exclusive_tools.join(","),
        reason.as_deref(),
    )
}

#[tauri::command]
pub fn list_resource_bindings() -> Vec<ResourceBindingRecord> {
    crate::database::list_resource_bindings()
}

#[tauri::command]
pub fn delete_resource_binding(extension_id: String) -> Result<(), String> {
    crate::database::delete_resource_binding(&extension_id)
}

#[tauri::command]
pub fn set_tool_resident(
    tool_id: String,
    extension_id: String,
    resident: bool,
) -> Result<(), String> {
    crate::services::tool_settings::ensure_tool_enabled(&tool_id)?;
    crate::database::set_tool_resident(&tool_id, &extension_id, resident)
}

#[tauri::command]
pub fn list_tool_residents(tool_id: String) -> Vec<String> {
    crate::database::list_tool_residents(&tool_id)
}

#[tauri::command]
pub fn list_presets() -> Vec<PresetRecord> {
    crate::database::list_presets()
}

#[tauri::command]
pub fn apply_preset(preset_id: String, tool_id: String) -> Result<PresetApplyResult, String> {
    // review F4：停用工具的预设写操作一律拒绝（W5 生效范围）
    crate::services::tool_settings::ensure_tool_enabled(&tool_id)?;
    let result = crate::services::preset::apply_preset(&preset_id, &tool_id)?;
    Ok(PresetApplyResult {
        success_count: result.success,
        failures: result.failures,
        conflicts: result.conflicts,
        stashed: result.stashed,
        disabled: result.disabled,
        restored_native: result.restored_native,
    })
}

#[tauri::command]
pub fn deactivate_preset(preset_id: String, tool_id: String) -> Result<(), String> {
    crate::services::tool_settings::ensure_tool_enabled(&tool_id)?;
    crate::services::preset::deactivate_preset(&preset_id, &tool_id)
}

#[tauri::command]
pub fn apply_preset_to_subagent(
    preset_id: String,
    tool_id: String,
    sub_agent_id: String,
) -> Result<PresetApplyResult, String> {
    crate::services::tool_settings::ensure_tool_enabled(&tool_id)?;
    let result =
        crate::services::preset::apply_preset_to_subagent(&preset_id, &tool_id, &sub_agent_id);
    Ok(PresetApplyResult {
        success_count: result.success,
        failures: result.failures,
        conflicts: result.conflicts,
        // 子 Agent 级应用不触发独占清扫/暂存（工具级专属语义），恒为空
        stashed: Vec::new(),
        disabled: Vec::new(),
        restored_native: Vec::new(),
    })
}

#[tauri::command]
pub fn deactivate_preset_from_subagent(
    preset_id: String,
    tool_id: String,
    sub_agent_id: String,
) -> Result<(), String> {
    crate::services::tool_settings::ensure_tool_enabled(&tool_id)?;
    crate::services::preset::deactivate_preset_from_subagent(&preset_id, &tool_id, &sub_agent_id)
}

/// 预设健康聚合（spec §13 检测侧收口）：不变量 + 未恢复暂存 + 账本-磁盘漂移三源合一
#[tauri::command]
pub fn get_preset_health() -> PresetHealth {
    crate::services::preset::preset_health()
}

/// 暂存条目回移（spec §13 体检卡片 ③；Task 15 裁决采方案 i）：
/// 按 id 在跨工具未恢复账目中查 entry（复用 unrestored_stash(None)，不新增 DAO），
/// 转发 stash::restore_stashed_skill；原位占用 / 暂存缺失等冲突原样报错
///（不覆盖现场），由前端 toast 透出
#[tauri::command]
pub fn restore_stash_entry(id: i64) -> Result<(), String> {
    let entry = crate::database::unrestored_stash(None)
        .into_iter()
        .find(|e| e.id == id)
        .ok_or_else(|| format!("暂存账本无此未恢复条目: id={}", id))?;
    crate::services::preset::stash::restore_stashed_skill(&entry)
}
